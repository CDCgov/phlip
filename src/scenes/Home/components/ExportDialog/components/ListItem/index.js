import React from 'react'
import PropTypes from 'prop-types'
import { default as MuiListItem } from '@material-ui/core/ListItem'
import Typography from '@material-ui/core/Typography'
import { Avatar, CircularLoader, FlexGrid, IconButton } from 'components'

/**
 * Wrapper for the list items in the export dialog / sections
 * @param props
 * @constructor
 */
export const ListItem = props => {
  const { inProgress, disabled, onExport, item, showAvatar, isSubItem } = props
  const listItemStyle = isSubItem
    ? { paddingTop: 8, paddingBottom: 8, fontSize: '.875rem' }
    : { padding: '12px 24px', fontSize: '1rem' }
  
  return (
    <MuiListItem style={listItemStyle}>
      <FlexGrid container type="row" align="center" style={{ width: '100%' }}>
        {showAvatar && <Avatar small userId={item.userId} {...item.avatarInfo} />}
        <Typography style={{ flex: '1 1 auto', fontSize: listItemStyle.fontSize, padding: isSubItem ? '0 16px' : 0 }}>
          {item.displayText}
        </Typography>
        {!inProgress &&
        <IconButton iconStyle={{ color: disabled ? 'unset' : 'black' }} disabled={disabled} onClick={onExport}>
          file_download
        </IconButton>}
        {inProgress && <CircularLoader type="indeterminate" style={{ width: 17, height: 17 }} />}
      </FlexGrid>
    </MuiListItem>
  )
}

ListItem.propTypes = {
  /**
   * Whether or not an export request is in progress for this item
   */
  inProgress: PropTypes.bool,
  /**
   * If item action should be disabled
   */
  disabled: PropTypes.bool,
  /**
   * Function to call when exporting for this item
   */
  onExport: PropTypes.func,
  /**
   * Item information
   */
  item: PropTypes.object,
  /**
   * Whether or not to show an avatar for this item
   */
  showAvatar: PropTypes.bool,
  /**
   * Whether it's a sub item
   */
  isSubItem: PropTypes.bool
}

ListItem.defaultProps = {
  isSubItem: true,
  showAvatar: true
}

export default ListItem
