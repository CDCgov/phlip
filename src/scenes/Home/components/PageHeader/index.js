import React from 'react'
import Typography from 'material-ui/Typography'
import PropTypes from 'prop-types'
import Button from 'components/Button'
import CircleIcon from 'components/CircleIcon'
import TextLink from 'components/TextLink'
import Container, { Column } from 'components/Layout'

export const PageHeader = ({ role }) => {
  return (
    <Container alignItems="center" style={{ height: '100px' }}>
      <Column flex>
        <Container alignItems="center" spacing={8}>
          <Column>
            <CircleIcon circleColor="error" iconColor="white" circleSize="35px" iconSize="24px">home</CircleIcon>
          </Column>
          <Column>
            <Typography type="headline">Project List</Typography>
          </Column>
        </Container>
      </Column>
      <Column>
        {role !== 'Coder' &&
        <TextLink to={{ pathname: '/project/add', state: { userDefined: null }}} style={{ color: 'white' }}>
          <Button value=" + Create new project" color="accent" />
        </TextLink>
        }
      </Column>
    </Container>
  )
}

PageHeader.propTypes = {
  role: PropTypes.string
}

export default PageHeader