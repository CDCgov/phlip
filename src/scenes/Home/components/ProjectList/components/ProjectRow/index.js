import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import TableRow from 'components/TableRow/index'
import Button from 'components/Button/index'
import TextLink from 'components/TextLink/index'
import IconButton from 'components/IconButton/index'
import TableCell from 'components/TableCell/index'
import * as actions from 'scenes/Home/actions'

const greyIcon = '#d4d4d4'

export const ProjectRow = ({ project, role, bookmarked, actions }) => {
  return (
      <TableRow key={project.id}>
        <TableCell key={`${project.id}-bookmarked`} style={{ width: 48 }}>
          <IconButton color={bookmarked ? '#fdc43b' : greyIcon} onClick={() => actions.toggleBookmark(project)}>
            {bookmarked ? 'bookmark' : 'bookmark_border'}
          </IconButton>
        </TableCell>
        <TableCell key={`${project.id}-name`} style={{ textAlign: 'left', width: 350 }}>
          <TextLink to={{ pathname: `/project/edit/${project.id}`, state: { projectDefined: { ...project }} }}>{project.name}</TextLink>
        </TableCell>
        <TableCell key={`${project.id}-dateLastEdited`}
                   style={{ color: '#9a9c9c', width: 150, maxWidth: 150, textAlign: 'unset' }} light>
          {new Date(project.dateLastEdited).toLocaleDateString()}
        </TableCell>
        <TableCell
          key={`${project.id}-lastEditedBy`}
          style={{ color: '#9a9c9c', fontStyle: 'italic', width: 150, maxWidth: 150, textAlign: 'unset' }}
          light
        >
          {project.lastEditedBy}
        </TableCell>
        {role !== 'Coder' &&
          <TableCell key={`${project.id}-protocol`} light>
            <TextLink to="/">Add/Edit</TextLink>
          </TableCell>
        }
        <TableCell key={`${project.id}-jurisdictions`} light>
          <TextLink to={`/project/${project.id}/jurisdictions`} id={project.id}>Add/Edit</TextLink>
        </TableCell>
        {role !== 'Coder' &&
        <TableCell key={`${project.id}-codingScheme`} light>
          <TextLink to="/">Edit</TextLink>
        </TableCell>
        }
        <TableCell key={`${project.id}-code`}>
          <Button raised={false} value="Code" listButton />
        </TableCell>
        {role !== 'Coder' &&
        <TableCell key={`${project.id}-validation`}>
          <Button raised={false} value="Validate" listButton />
        </TableCell>
        }
        <TableCell key={`${project.id}-export`}>
          <IconButton color={greyIcon} onClick={() => actions.onExport}>file_download</IconButton>
        </TableCell>
      </TableRow>
  )
}

ProjectRow.propTypes = {
}

const mapStateToProps = (state, ownProps) => ({
  project: state.scenes.home.main.projects.byId[ownProps.id],
  role: state.data.user.currentUser.role,
  bookmarked: state.scenes.home.main.bookmarkList.includes(ownProps.id)
})

const mapDispatchToProps = (dispatch) => ({ actions: bindActionCreators(actions, dispatch) })

export default connect(mapStateToProps, mapDispatchToProps)(ProjectRow)
