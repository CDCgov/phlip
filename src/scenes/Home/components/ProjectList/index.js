import React from 'react'
import PropTypes from 'prop-types'
import { TableBody, TableHead, TableFooter, TableRow } from 'material-ui/Table'
import Container, { Column } from 'components/Layout'
import Card from 'components/Card'
import Table from 'components/Table'
import ProjectTableRow from './components/ProjectRow'
import ProjectTableHead from './components/ProjectTableHead'
import TablePagination from 'components/TablePagination'

export const ProjectList = props => {
  const { projectIds, user, page, rowsPerPage, projectCount, sortBy, direction, sortBookmarked } = props
  const { handlePageChange, handleRowsChange, handleRequestSort, handleSortBookmarked } = props
  return (
    <Container flex>
      <Column flex displayFlex style={{ overflowX: 'auto' }} component={<Card/>}>
        <Table style={{ borderCollapse: 'separate' }}>
          <TableHead>
            <ProjectTableHead
              role={user.role}
              sortBy={sortBy}
              direction={direction}
              sortBookmarked={sortBookmarked}
              onRequestSort={handleRequestSort}
              onSortBookmarked={handleSortBookmarked}
            />
          </TableHead>
          <TableBody>
            {projectIds.map(id => (
              <ProjectTableRow key={id} id={id} />
            ))}
          </TableBody>
        </Table>
        <div style={{ display: 'flex', flex: 1, paddingBottom: '50px' }}/>
        <Table>
          <TableFooter>
            <TableRow>
              <TablePagination
                count={projectCount}
                rowsPerPage={rowsPerPage}
                page={page}
                onChangePage={(event, page) => handlePageChange(page)}
                onChangeRowsPerPage={(event) => handleRowsChange(event.target.value)}
              />
            </TableRow>
          </TableFooter>
        </Table>
      </Column>
    </Container>
  )
}


export default ProjectList
