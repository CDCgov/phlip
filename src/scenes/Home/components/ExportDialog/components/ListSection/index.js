import React from 'react'
import PropTypes from 'prop-types'
import { Icon, IconButton } from 'components'
import List from '@material-ui/core/List'
import { default as MuiListItem } from '@material-ui/core/ListItem'
import Collapse from '@material-ui/core/Collapse'
import ListItemText from '@material-ui/core/ListItemText'
import ListItem from '../ListItem'

/**
 * Renders the expandable part for a export dialog list section
 * @param props
 * @returns {*}
 * @constructor
 */
export const ListSection = props => {
  const { users, section, expanded, sectionText, onExport, onExpand, inProgress } = props
  const validatorUser = {
    avatarInfo: {
      initials: <Icon size={20} color="white" style={{ fontWeight: 800 }}>check</Icon>,
      style: { backgroundColor: '#80d134' },
      avatar: ''
    },
    userId: null,
    firstName: 'Validated',
    lastName: 'Codes'
  }
  const allUsers = [...users, validatorUser]
  
  return (
    <>
      <MuiListItem onClick={onExpand(section)} selected={expanded}>
        <ListItemText primary={sectionText} />
        <IconButton iconStyle={{ color: 'black' }}>
          {expanded ? 'expand_less' : 'expand_more'}
        </IconButton>
      </MuiListItem>
      <Collapse in={expanded}>
        <List component="div" disablePadding>
          {allUsers.map(user => {
            const userInProgress = inProgress.type !== section
              ? false
              : user.userId === null
                ? inProgress.user === 'val'
                : inProgress.user === user.userId
            
            return (
              <ListItem
                key={`${section}-${user.userId}`}
                isUser={user.userId !== null}
                item={{ ...user, displayText: `${user.firstName} ${user.lastName}` }}
                inProgress={userInProgress}
                disabled={inProgress.type !== null && !userInProgress}
                onExport={onExport(section, user.userId === null ? null : user)}
              />
            )
          })}
        </List>
      </Collapse>
    </>
  )
}

ListSection.propTypes = {
  /**
   * Is this section expanded
   */
  expanded: PropTypes.bool,
  /**
   * List of users available for export
   */
  users: PropTypes.array,
  /**
   * String name of section (type of export)
   */
  section: PropTypes.string,
  /**
   * Section list header
   */
  sectionText: PropTypes.string,
  /**
   * User clicked the download button for a user or validator
   */
  onExport: PropTypes.func,
  /**
   * Expand or collapse section
   */
  onExpand: PropTypes.func,
  /**
   * User in progress for export if any
   */
  inProgress: PropTypes.object
}

export default ListSection
