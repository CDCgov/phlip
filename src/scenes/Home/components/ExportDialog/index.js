import React from 'react'
import PropTypes from 'prop-types'
import Modal, { ModalTitle, ModalContent, ModalActions } from 'components/Modal'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import ListItemIcon from '@material-ui/core/ListItemIcon'
import Divider from '@material-ui/core/Divider'
import Icon from 'components/Icon'

export const ExportDialog = ({ onClose, onChooseExport, open }) => {
  const actions = [
    { value: 'Cancel', onClick: onClose, type: 'button', otherProps: { 'aria-label': 'Close modal' } }
  ]

  return (
    <Modal open={open} onClose={onClose}>
      <ModalTitle title="Choose export type" />
      <Divider />
      <ModalContent style={{ minWidth: 400, padding: 0 }}>
        <List>
          <ListItem button onClick={() => onChooseExport('numeric')} style={{ paddingLeft: 25, paddingRight: 25 }}>
            <ListItemIcon><Icon style={{ color: 'black' }}>file_download</Icon></ListItemIcon>
            <ListItemText primary="Coded Data - Numeric" />
          </ListItem>
          <ListItem button onClick={() => onChooseExport('text')} style={{ paddingLeft: 25, paddingRight: 25 }}>
            <ListItemIcon><Icon style={{ color: 'black' }}>file_download</Icon></ListItemIcon>
            <ListItemText primary="Coded Data - Text" />
          </ListItem>
          <ListItem button onClick={() => onChooseExport('codebook')} style={{ paddingLeft: 25, paddingRight: 25 }}>
            <ListItemIcon><Icon style={{ color: 'black' }}>file_download</Icon></ListItemIcon>
            <ListItemText primary="Codebook" />
          </ListItem>
        </List>
      </ModalContent>
      <ModalActions actions={actions} />
    </Modal>
  )
}

ExportDialog.propTypes = {
  /**
   * Function to call when the modal is closed
   */
  onClose: PropTypes.func,
  /**
   * Function to call when the user chooses an export type
   */
  onChooseExport: PropTypes.func,
  /**
   * Whether or not the modal is open
   */
  open: PropTypes.bool
}

export default ExportDialog