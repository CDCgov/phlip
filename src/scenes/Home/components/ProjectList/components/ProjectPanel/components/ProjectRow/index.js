import React from 'react'
import PropTypes from 'prop-types'
import { FlexGrid, IconButton, Icon } from 'components'
import moment from 'moment'
import theme from 'services/theme'

export const ProjectRow = props => {
  const { project, bookmarked, toggleBookmark, handleExpandProject } = props
  const date = moment.utc(project.dateLastEdited).local().format('M/D/YYYY')
  const greyIcon = theme.palette.greyText
  const isLocked = project.status === 2

  return (
    <FlexGrid type="row" container flex style={{ minHeight: 60 }} onClick={handleExpandProject} id={project.name.replace(' ','_')}>
      <FlexGrid container type="row" align="center" padding="0 0 0 24px">
        <IconButton
          color={bookmarked ? '#fdc43b' : greyIcon}
          onClick={event => {
            event.stopPropagation()
            toggleBookmark(project)
          }}
          disabled={isLocked}
          tooltipText="Bookmark project"
          aria-label="Bookmark this project"
          id={`bookmark-project-${project.id}`}>
          {bookmarked ? 'bookmark' : 'bookmark_border'}
        </IconButton>
      </FlexGrid>
      <FlexGrid padding="0 12px" container type="row" align="center" flex style={{ flexBasis: '32%' }}>
        <span>{project.name}</span>
        {isLocked && <Icon size={18} color="#757575" style={{ marginLeft: 5 }}>lock</Icon>}
      </FlexGrid>
      <FlexGrid
        container
        type="row"
        align="center"
        justify="flex-end"
        padding="0 24px 0 0"
        flex
        style={{ flexBasis: '10%' }}>
        <FlexGrid style={{ width: '45%' }}>
          <span style={{ color: 'black' }}>Date Last Edited:{' '}</span>
          <span>{date}</span>
        </FlexGrid>
        <span style={{ width: '8%' }} />
        <FlexGrid style={{ width: '45%' }}>
          <span style={{ color: 'black' }}>Last Edited By:{' '}</span>
          {project.lastEditedBy}
        </FlexGrid>
      </FlexGrid>
    </FlexGrid>
  )
}

ProjectRow.propTypes = {
  project: PropTypes.object,
  bookmarked: PropTypes.bool,
  isCoder: PropTypes.bool,
  toggleBookmark: PropTypes.func,
  handleExpandProject: PropTypes.func
}

export default ProjectRow
