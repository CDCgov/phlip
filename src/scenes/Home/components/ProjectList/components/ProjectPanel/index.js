import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import actions from 'scenes/Home/actions'
import moment from 'moment'
import GridList from '@material-ui/core/GridList'
import Typography from '@material-ui/core/Typography'
import Collapse from '@material-ui/core/Collapse'
import { FlexGrid, IconButton, Link, Button, GridListTile } from 'components'
import { FileDocument, City, FormatListBulleted, ClipboardCheckOutline, FileExport } from 'mdi-material-ui'
import ProjectRow from './components/ProjectRow'
import silhouette from './silhouette.png'
import theme from 'services/theme'

export class ProjectPanel extends Component {
  static propTypes = {
    project: PropTypes.object,
    actions: PropTypes.object,
    role: PropTypes.string,
    expanded: PropTypes.bool,
    users: PropTypes.array,
    allUsers: PropTypes.object,
    bookmarked: PropTypes.bool,
    onExport: PropTypes.func,
    length: PropTypes.number,
    index: PropTypes.number,
    handleExpandProject: PropTypes.func,
    handleEditProject: PropTypes.func
  }
  
  /**
   * Handles when a project is clicked to open
   * @param event
   */
  handleChange = event => {
    this.props.handleExpandProject(this.props.project.id, event)
    this.props.actions.getProjectUsers(this.props.project.id, {
      id: this.props.project.createdById,
      email: this.props.project.createdByEmail
    })
  }
  
  /**
   * Determines the grid size for avatars
   * @returns {{square: number, size: number}}
   */
  determineGridSize = () => {
    let size = 0, square = 0, i = 0
    const { users } = this.props
    
    while (size === 0) {
      const sq = i ** 2
      if (users.length <= sq) {
        square = sq
        size = i
      }
      i++
    }
    
    return { square, size }
  }
  
  /**
   * Gets users for a project and fills in empty slots in the grid
   * @param square
   * @returns {...ProjectPanel.props.users[]}
   */
  populateUsers = square => {
    const { users, project } = this.props
    if (square === users.length) {
      return [...users]
    }
    
    const userTiles = [...users]
    const blanks = Array.from({ length: square - users.length }, (v, i) => ({
      id: `${project.id}-blank-${i}`,
      blank: true
    }))
    return userTiles.concat(blanks)
  }
  
  /**
   * Handles when a user clicks the export button
   */
  onClickExport = () => {
    this.props.onExport(this.props.project)
  }
  
  render() {
    const { project, role, bookmarked, actions, index, length, users, allUsers, expanded, handleEditProject } = this.props
    
    const isCoder = role === 'Coder'
    const greyIcon = theme.palette.greyText
    const iconStyle = { fontSize: 18, paddingLeft: 5 }
    const listingStyle = { fontSize: 14, fontWeight: 500, color: '#7b7b7b' }
    const date = moment.utc(project.dateLastEdited).local().format('M/D/YYYY')
    const createdDate = moment.utc(project.dateCreated).local().format('M/D/YYYY')
    
    const { square, size } = this.determineGridSize()
    const userData = expanded ? this.populateUsers(square) : []
    const avatarCols = size, avatarRows = size, cellHeight = 300 / size
    
    const rowStyles = {
      boxShadow: '0px 1px 3px 0px rgba(0, 0, 0, 0.2), 0px 1px 1px 0px rgba(0, 0, 0, 0.14), 0px 2px 1px -1px rgba(0, 0, 0, 0.12)',
      position: 'relative',
      backgroundColor: 'white',
      fontSize: '0.8125rem',
      borderRadius: expanded ? 2 : 0,
      overflow: 'hidden'
    }
    
    const collapseStyles = {
      transitionDuration: '150ms',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      width: '100%'
    }
    
    const containerStyles = {
      paddingBottom: ((index === 0 && length > 1) || index < length - 1)
        ? expanded
          ? 30
          : 0
        : 0,
      paddingTop: expanded
        ? index > 0
          ? 30
          : 0
        : 0
    }
    
    const panelButtonProps = {
      style: { marginLeft: 10 },
      color: 'white',
      textColor: 'black',
      disableRipple: true
    }
    
    return (
      <>
        {containerStyles.paddingTop === 30 && <div style={{ height: 30 }} onClick={this.handleChange} />}
        <FlexGrid>
          <FlexGrid container justify="center" style={rowStyles} flex>
            {!expanded && <ProjectRow
              isCoder={isCoder}
              project={project}
              bookmarked={bookmarked}
              toggleBookmark={actions.toggleBookmark}
              handleExpandProject={this.handleChange}
            />}
            <Collapse in={expanded} style={collapseStyles}>
              <FlexGrid container type="row" style={{ width: '100%', overflow: 'auto' }}>
                <FlexGrid container type="column" style={{ width: 300, backgroundColor: 'white' }}>
                  <GridList
                    style={{ width: 300, margin: 0 }}
                    cellHeight={cellHeight}
                    spacing={2}
                    cols={avatarCols}
                    rows={avatarRows}>
                    {userData.map(oneCoder => {
                      const coder = oneCoder.blank
                        ? oneCoder
                        : allUsers[oneCoder.userId] !== undefined
                          ? allUsers[oneCoder.userId]
                          : oneCoder
  
                      return (
                        <GridListTile key={coder.id || coder.userId}>
                          {!coder.blank ? (
                            (coder.avatar === '' || coder.avatar === undefined)
                              ? (
                                <FlexGrid
                                  container
                                  type="row"
                                  justify="center"
                                  align="flex-end"
                                  title={coder.username}
                                  style={{ backgroundColor: '#f9f9f9' }}>
                                  <img
                                    src={silhouette}
                                    alt={`${coder.username}'s avatar`}
                                    style={{ width: '100%', height: cellHeight }}
                                  />
                                  <Typography
                                    style={{
                                      fontWeight: 300,
                                      fontSize: '1.1rem',
                                      color: 'white',
                                      backgroundColor: `rgb(${144},${141},${141})`,
                                      width: '100%',
                                      textAlign: 'center',
                                      height: 25,
                                      paddingLeft: 3,
                                      position: 'absolute'
                                    }}>
                                    {coder.initials}
                                  </Typography>
                                </FlexGrid>
                              ) : (
                                <img
                                  alt={`avatar-${coder.lastName}`}
                                  src={coder.avatar}
                                  title={coder.username}
                                />
                              )
                          ) : <div style={{ backgroundColor: '#f9f9f9', width: '100%', height: cellHeight }} />}
                        </GridListTile>
                      )
                    })}
                  </GridList>
                </FlexGrid>
                <FlexGrid container type="column" flex>
                  <FlexGrid
                    container
                    type="row"
                    align="center"
                    padding={30}
                    onClick={this.handleChange}
                    style={{ backgroundColor: '#f9f9f9' }}>
                    <FlexGrid
                      container
                      type="row"
                      flex
                      align="center"
                      justify="flex-start"
                      style={{ fontSize: 20, minWidth: 150 }}>
                      <IconButton
                        color={bookmarked ? '#fdc43b' : greyIcon}
                        onClick={() => actions.toggleBookmark(project)}
                        tooltipText="Bookmark project"
                        aria-label="Bookmark this project"
                        id={`bookmark-project-${project.id}`}>
                        {bookmarked ? 'bookmark' : 'bookmark_border'}
                      </IconButton>
                      <FlexGrid style={{ width: 20 }} />
                      <Typography variant="title" style={{ marginRight: 5 }}>{project.name}</Typography>
                      {!isCoder && <IconButton
                        color="secondary"
                        iconSize={20}
                        id={`edit-project-${project.id}`}
                        tooltipText="Edit Project Details"
                        aria-label={`Edit Project ${project.name}`}
                        onClick={handleEditProject(project)}>
                        edit
                      </IconButton>}
                    </FlexGrid>
                    <FlexGrid container type="row" flex justify="flex-end" align="stretch" style={{ height: 40 }}>
                      {!isCoder && (<Button
                        id={`${project.id}-edit-jurisdictions`}
                        to={{ pathname: `/project/${project.id}/jurisdictions`, state: { modal: true } }}
                        aria-label="Add and edit project jurisdictions"
                        component={Link}
                        {...panelButtonProps}>
                        Jurisdictions
                        <City style={iconStyle} />
                      </Button>)}
                      <Button
                        component={Link}
                        aria-label="Documents in this project"
                        to={{ pathname: `/docs`, state: { projectDefined: true, project } }}
                        {...panelButtonProps}>
                        Documents
                        <FileDocument style={iconStyle} />
                      </Button>
                      {!isCoder && (<Button
                        component={Link}
                        aria-label="Add and edit project coding scheme"
                        to={`/project/${project.id}/coding-scheme`}
                        {...panelButtonProps}>
                        Coding Scheme
                        <FormatListBulleted style={iconStyle} />
                      </Button>)}
                      <Button
                        component={Link}
                        aria-label="Add and edit project protocol"
                        to={`/project/${project.id}/protocol`}
                        {...panelButtonProps}>
                        Protocol
                        <ClipboardCheckOutline style={iconStyle} />
                      </Button>
                      <Button
                        aria-label="Export validated questions"
                        onClick={this.onClickExport}
                        id={`export-validated-${project.id}`}
                        {...panelButtonProps}>
                        Export
                        <FileExport style={iconStyle} />
                      </Button>
                    </FlexGrid>
                  </FlexGrid>
                  <FlexGrid container type="row">
                    <FlexGrid container flex padding={30}>
                      <Typography variant="body1" style={{ paddingBottom: 10 }}>
                        <span style={listingStyle}>Created by:{' '}</span>
                        {project.createdBy} -{' '}
                        <span style={{ fontSize: `0.8rem`, color: '#7b7b7b' }}>{createdDate}</span>
                      </Typography>
                      <Typography variant="body1" style={{ paddingBottom: 10 }}>
                        <span style={listingStyle}>Last edited by:{' '}</span>
                        {project.lastEditedBy} -{' '}
                        <span style={{ fontSize: `0.8rem`, color: '#7b7b7b' }}>{date}</span>
                      </Typography>
                      <Typography variant="body1" style={{ paddingBottom: 10 }}>
                        <span style={listingStyle}>Coordinator(s):{' '}</span>
                        <span>
                          {users.filter((oneCoder) => {
                            return oneCoder.role !== 'Coder'
                          }).map((oneCoder) => {
                            return oneCoder.firstName + ' ' + oneCoder.lastName
                          }).join(', ')}
                        </span>
                      </Typography>
                      <Typography variant="body1" style={{ paddingBottom: 10 }}>
                        <span style={listingStyle}>Coder(s):{' '}</span>
                        <span>
                          {users.filter((oneCoder) => {
                            return oneCoder.role === 'Coder'
                          }).map((oneCoder) => {
                            return oneCoder.firstName + ' ' + oneCoder.lastName
                          }).join(', ')}
                        </span>
                      </Typography>
                      <Typography variant="body1">
                        <span style={listingStyle}>Type:{' '}</span>
                        <span>{(project.type === 1) ? 'Legal Scan' : 'Policy Surveillance'}</span>
                      </Typography>
                      <FlexGrid />
                    </FlexGrid>
                    <FlexGrid container type="row" align="flex-end" flex padding={30} justify="flex-end">
                      <Button
                        raised={false}
                        value="Code"
                        listButton
                        aria-label="Code project"
                        component={Link}
                        to={{ pathname: `/project/${project.id}/code` }}
                        style={{ width: 150, height: 50, marginRight: 20, borderRadius: 3 }}
                      />
                      {!isCoder &&
                      <Button
                        raised={false}
                        value="Validate"
                        listButton
                        aria-label="Validate project"
                        component={Link}
                        to={{ pathname: `/project/${project.id}/validate` }}
                        style={{ width: 150, height: 50, borderRadius: 3 }}
                      />}
                    </FlexGrid>
                  </FlexGrid>
                </FlexGrid>
              </FlexGrid>
            </Collapse>
          </FlexGrid>
        </FlexGrid>
        {containerStyles.paddingBottom === 30 && <div style={{ height: 30 }} onClick={this.handleChange} />}
      </>
    )
  }
}

/* istanbul ignore next */
const mapStateToProps = (state, ownProps) => {
  const homeState = state.scenes.home.main
  const p = state.data.projects.byId[ownProps.id]
  return {
    project: p,
    role: state.data.user.currentUser.role,
    bookmarked: homeState.bookmarkList.includes(ownProps.id),
    users: p.projectUsers,
    allUsers: state.data.user.byId
  }
}

/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({ actions: bindActionCreators(actions, dispatch) })

export default connect(mapStateToProps, mapDispatchToProps)(ProjectPanel)
