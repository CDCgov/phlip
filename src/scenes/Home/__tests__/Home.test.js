import React from 'react'
import { shallow, mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { MuiThemeProvider } from 'material-ui/styles'
import theme from 'services/theme'
import { Home } from '../index'
import { ProjectList } from '../components/ProjectList'
import { mockProjects } from 'data/mockProjects'
import { PageHeader } from 'components/PageHeader'

const props = {
  actions: {
    getProjectsRequest: jest.fn(),
    sortProjects: jest.fn(),
    updatePage: jest.fn(),
    updateRows: jest.fn(),
    toggleBookmark: jest.fn(),
    sortBookmarked: jest.fn(),
    updateSearchValue: jest.fn()
  },
  user: {
    id: 2,
    firstName: 'Kristin',
    lastName: 'Muterspaw',
    email: 'kmuterspaw@cdc.gov',
    password: 'test',
    role: 'Coordinator'
  },
  error: false,
  errorContent: '',
  projects: mockProjects,
  visibleProjects: [],
  bookmarkList: [],
  sortBookmarked: false,
  sortBy: 'dateLastEdited',
  direction: 'desc',
  page: 0,
  rowsPerPage: '10',
  searchValue: ''
}

const setup = (otherProps = {}, initialEntries = ['/']) => {
  return mount(
    <MemoryRouter initialEntries={initialEntries}>
      <MuiThemeProvider theme={theme}>
        <Home {...props} {...otherProps} />
      </MuiThemeProvider>
    </MemoryRouter>
  )
}

describe('Home scene', () => {
  test('should render correctly', () => {
    expect(shallow(<Home {...props} />)).toMatchSnapshot()
  })

  test('should render ProjectList and PageHeader components', () => {
    const wrapper = setup()
    expect(wrapper.find(ProjectList)).toHaveLength(1)
  })

  test('should open the New Project modal when Create New Project is clicked', () => {
    const wrapper = setup()
    wrapper.find(PageHeader).find('Button').at(0).simulate('click')
    wrapper.update()
    expect(wrapper.find('Modal')).toHaveLength(1)
  })

  describe('Error handling', () => {
    test('should display an error message if prop: error is true', () => {
      const wrapper = setup({ error: true, errorContent: 'We could not get projects.' })
      expect(wrapper.find('CardError')).toHaveLength(1)
      expect(wrapper.find('ProjectList')).toHaveLength(0)
    })

    test('should display the content of errorContent prop in error message', () => {
      const wrapper = setup({ error: true, errorContent: 'We could not get projects.' })
      expect(wrapper.find('CardError').text()).stringMatching('Uh-oh, something went wrong. We could not get projects.')
    })
  })

  describe('Unauthorized routes', () => {
    test('should display PageNotFound if the user role is Coder and they try to go to add project', () => {
      const wrapper = setup({ user: { role: 'Coder' }, location: { pathname: '/project/add' } }, ['/project/add'])
      expect(wrapper.find('PageNotFound')).toHaveLength(1)
    })

    test('should display PageNotFound if the user role is Coder and they try to go to jurisdictions', () => {
      const wrapper = setup({
        user: { role: 'Coder' },
        location: { pathname: '/project/1/jurisdictions' }
      }, ['/project/1/jurisdictions'])
      expect(wrapper.find('PageNotFound')).toHaveLength(1)
    })
  })
})
