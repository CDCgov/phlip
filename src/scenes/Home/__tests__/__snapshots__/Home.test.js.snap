// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home scene should render correctly 1`] = `
<Container
  column={true}
  flex={true}
  spacing={0}
>
  <PageHeader
    role="Coordinator"
  />
  <withStyles(Divider) />
  <SearchBar
    handleSearchValueChange={[Function]}
    height={57}
    searchValue=""
  />
  <ProjectList
    direction="desc"
    handlePageChange={[Function]}
    handleRequestSort={[Function]}
    handleRowsChange={[Function]}
    handleSortBookmarked={[Function]}
    page={0}
    projectIds={Array []}
    rowsPerPage={10}
    sortBookmarked={false}
    sortBy="dateLastEdited"
    user={
      Object {
        "email": "kmuterspaw@cdc.gov",
        "firstName": "Kristin",
        "id": 2,
        "lastName": "Muterspaw",
        "password": "test",
        "role": "Coordinator",
      }
    }
  />
  <Route
    component={[Function]}
    path="/new/project"
  />
</Container>
`;
